### DISCUSSION TITLE
How to achieve UE registration and PDU procedures using Packet Rusher

### DISCUSSION THREAD

**Discussion by kirankumar-Athirala** [HewlettPackard/PacketRusher#136] [Q&A]
Hello

I am planning to use packet rusher  that involves triggering 5G core procedures such as registration, PDU connection, deregistration, and handover using Packet Rusher. Specifically, I need to:

Trigger 100 registration instances.
Trigger 30 PDU connection instances.
Trigger 50 deregistration instances.
Trigger 20 combined registration and PDU connection instances.
Trigger 10 handover procedures.
Amf load testing

For my setup, I am using the Radiant Open5G Core Helm chart to deploy Open5GCore on Kubernetes and also using the PacketRusher Helm chart to deploy Packet Rusher. Could you please provide guidance or examples on how to achieve these using Packet Rusher? Any documentation and  detailed steps would be greatly appreciated.

Thank you!

**Comment by kirankumar-Athirala** [comment]
Hi Good evening,

I am deploying PacketRusher in Kubernetes and successfully launching it with a single UE. However, when I run PacketRusher with the command ./packetrusher --config config.yml amf-load-loop -n 20 -t 20, it crashes. You can find the crash log in the provided file.
[packet_rusher_amf_load_test.txt](https://github.com/user-attachments/files/16365151/packet_rusher_amf_load_test.txt)


Additionally, I attempted to launch multiple UEs using ./packetrusher --config config.yml multi-ue -n 1 -d true -t true, but I encountered the following issue:
PDU address received: 10.45.0.3
time="2024-07-24T15:01:22Z" level=info msg="[UE][GTP] Interface for UE 0000000001 has not been created. Tunnel has been disabled."

Even when I tried launching with a single UE in multi-ue, I still faced the same problem. Could you please help me resolve this issue?
[packet_rusher_multi_ue.txt](https://github.com/user-attachments/files/16365163/packet_rusher_multi_ue.txt)

I also tried running packet rusher with --multi-ue-pdu -n 5 --tr 60000 --td 60000 --npdu 1 -l true . but I didn't observer de registration and re registration of UE's
 
[packet_rusher_multi_ue_pdu.txt](https://github.com/user-attachments/files/16365300/packet_rusher_multi_ue_pdu.txt)

could you please help me to debug these issues. Thanks in advance

Best Regards
Kiran

**Comment by linouxis9** [comment]
Hi @kirankumar-Athirala,

I'm happy to help!

To see all the possible commands, I would advise you to run the following the command: `sudo ./packetrusher multi-ue --help`.
One example is for eg. `sudo ./packetrusher multi-ue -n 1 -xnh 3000 -idl 13000 -ngh 23000`.
This command will create and register an UE, make it do a Xn Handover to another UE after 3000 ms, then a switch to CM-IDLE and then CM-CONNECTED procedure after 13000 ms, and then finally a NGAP Handover after 23000 ms.

For the Handover procedures, you need two gNodeBs. As such, PacketRusher automatically creates a second gNodeB with the same configuration as the first gNodeB, but with the IP used by the second gNodeB next to the first gNodeB: if the first gNodeB has IP 10.0.0.1 in config.yml, the second gNodeB created automatically by PacketRusher will have IP 10.0.0.2 and this cannot be configured differently for now.
As such, make sure to add, for example, the second IP 10.0.0.2 to your network interface (eg. sudo ip addr add eth0 10.0.0.2).

What you are missing is the --tunnel and --dedicatedGnb parameters which are by default disabled when using multi-ue, eg: `sudo ./packetrusher multi-ue -n 1`.

Please note, that you need a gNodeB per UE when using --tunnel and --dedicatedGnb.
More information on --dedicatedGnb here: https://github.com/HewlettPackard/PacketRusher/issues/65#issuecomment-1895554491 and on why this limitation exists for now here: https://github.com/HewlettPackard/PacketRusher/issues/111#issuecomment-2049126443

I hope it helped, feel free to reach out if you need more help!

Thanks and best regards,
Valentin
